src/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/Marshal/GetTypedObjectForIUnknownTests.cs

```diff
@@ -73,38 +124,86 @@ public static IEnumerable<object[]> GetTypedObjectForIUnknown_Invalid_TestData()
+                AssertExtensions.Throws<ArgumentException>("t", () => Marshal.GetTypedObjectForIUnknown(ptr, type));
             }
             finally
             {
-                Marshal.Release(iUnknown);
+                Marshal.Release(ptr);
+            }
+        }
+
+        public static IEnumerable<object[]> GetTypedObjectForIUnknownType_UncastableObject_TestData()
+        {
+            yield return new object[] { new object(), typeof(AbstractClass) };
+            yield return new object[] { new object(), typeof(NonGenericClass) };
+            yield return new object[] { new object(), typeof(NonGenericStruct) };
+            yield return new object[] { new object(), typeof(NonGenericStruct) };
+            yield return new object[] { new object(), typeof(NonGenericInterface) };
+
+            yield return new object[] { new NonGenericClass(), typeof(IFormattable) };
+            yield return new object[] { new ClassWithInterface(), typeof(IFormattable) };
+
+            yield return new object[] { new object(), typeof(int).MakePointerType() };
+
+            AssemblyBuilder assemblyBuilder = AssemblyBuilder.DefineDynamicAssembly(new AssemblyName("Assembly"), AssemblyBuilderAccess.RunAndCollect);
+            ModuleBuilder moduleBuilder = assemblyBuilder.DefineDynamicModule("Module");
+            TypeBuilder typeBuilder = moduleBuilder.DefineType("Type");
+            Type collectibleType = typeBuilder.CreateType();
+            yield return new object[] { new object(), collectibleType };
+        }
+
+        [Theory]
+        [MemberData(nameof(GetTypedObjectForIUnknownType_UncastableObject_TestData))]
+        [PlatformSpecific(TestPlatforms.Windows)]
+        public void GetTypedObjectForIUnknown_UncastableObject_ThrowsInvalidCastException(object o, Type type)
+        {
+            IntPtr ptr = Marshal.GetIUnknownForObject(o);
+            try
+            {
+                Assert.Throws<InvalidCastException>(() => Marshal.GetTypedObjectForIUnknown(ptr, type));
+            }
+            finally
+            {
+                Marshal.Release(ptr);
             }
         }
 
+        public static IEnumerable<object[]> GetTypedObjectForIUnknown_ArrayObjects_TestData()
+        {
+            yield return new object[] { new int[] { 10 } };
+            yield return new object[] { new int[][] { new int[] { 10 } } };
+            yield return new object[] { new int[,] { { 10 } } };
+        }
+
         [Theory]
+        [MemberData(nameof(GetTypedObjectForIUnknown_ArrayObjects_TestData))]
         [PlatformSpecific(TestPlatforms.Windows)]
-        [InlineData(typeof(AbstractClass))]
-        [InlineData(typeof(NonGenericClass))]
-        [InlineData(typeof(NonGenericStruct))]
-        [InlineData(typeof(NonGenericInterface))]
-        public void GetTypedObjectForIUnknown_UncastableType_ThrowsInvalidCastException(Type type)
+        public void GetTypedObjectForIUnknown_ArrayType_ThrowsBadImageFormatException(object o)
```

Comment:
Could you create a new issue for this case?

Reply:
What do you mean? The test passes on both .NET Framework and .NET Core. Do you mean making an issue for the fact it throws a BIFE?

Reply:
>>Do you mean making an issue for the fact it throws a BIFE? 

Yes, please